File: bin/foreign-regexp-replace-aux.pl
#!/usr/bin/env perl
# -*- coding: utf-8-unix -*-
use strict;
use warnings;
use 5.008;

use Encode;
use utf8;

package main;
use English qw( -no_match_vars );
use FileHandle;

sub interpolate_fn_gen {
    # Interpolate replacement string in environment
    # which has no lexical variable.
    #
    # Special-variables in the replacement string
    # will be interpolated.
    eval 'sub {"'. escape_str_for_interpolate_fn_gen(${$_[0]}) .'"}';
}

sub eval_fn_gen {
    # Eval replacement string in environment
    # which has no lexical variable.
    eval 'sub {'.${$_[0]}.'}';
}

sub escape_str_for_interpolate_fn_gen {
    my $txt = shift;
    $txt =~ s/"/\\"/og;
    $txt
}

sub escape_perl_str_for_emacs {
    my $r_txt = shift;
    ${$r_txt} =~ s/\\/\\\\/og;
    ${$r_txt} =~ s/"/\\"/og;
}

sub process_replace {
    my $r_str_body   = shift;
    my $r_str_regx   = shift;
    my $r_str_rpla   = shift;
    my $dot_p        = shift;
    my $case_p       = shift;
    my $ext_p        = shift;
    my $eval_p       = shift;
    my $limit        = shift;
    my $pos_start    = shift;
    my $rgn_beg      = shift;
    my $rgn_end      = shift;
    
    my $pos_wrap_end = undef;
    my $count        = 0;
    
    my $regx = eval ("qr/\${\$r_str_regx}/mo" .
                     ( $dot_p  ? "s" : "") .
                     (!$case_p ? "i" : "") .
                     ( $ext_p  ? "x" : ""));
    die $EVAL_ERROR if $EVAL_ERROR;
    
    my $interpolate_fn = ($eval_p
                          ? eval_fn_gen($r_str_rpla)
                          : interpolate_fn_gen($r_str_rpla));
    die "Syntax error in replacement \"${$r_str_rpla}\":\n${EVAL_ERROR}" if $EVAL_ERROR;
    
    my $replace_fn = sub {
        my $rgn_beg = shift;
        my $rgn_end = shift;
        my $wrap_p  = shift;
        
        
        pos(${$r_str_body}) = $rgn_beg;
        
        while (((defined $limit) ? ($count < $limit) : 1) && (${$r_str_body} =~ m/${regx}/g)) {
            my $match_beg = $LAST_MATCH_START[0];
            my $match_end = $LAST_MATCH_END  [0];
            
            last if ($match_end > $rgn_end);
            last if ($wrap_p && (defined $pos_wrap_end) && ($pos_wrap_end <= $match_beg));
            $pos_wrap_end = $match_beg if ((not $wrap_p) && (not (defined $pos_wrap_end)));
            
            my $replacement = eval { $interpolate_fn->() };
            die "Error while interpolating replacement \"${$r_str_rpla}\":\n${EVAL_ERROR}" if $EVAL_ERROR;
            
            escape_perl_str_for_emacs(\$replacement);
            
            print " ((";
            print $match_beg, ' ';
            print $match_end, ' ';
            foreach my $i (1 .. $#LAST_MATCH_START) {
                print $LAST_MATCH_START[$i], ' ';
                print $LAST_MATCH_END  [$i], ' ';
            }
            print " )";
            print '"', $replacement, '"';
            print " )", "\n";
            
            ++$count;
        }
    };
    
    $rgn_beg   = $rgn_beg || 0;
    $rgn_end   = $rgn_end || length(${$r_str_body});
    $pos_start = (($pos_start < $rgn_beg)
                  ? $rgn_beg
                  : (($pos_start > $rgn_end)
                     ? $rgn_end
                     : $pos_start));
    
    print "(setq result '(";
    print " (";
    $replace_fn->($pos_start, $rgn_end, 0);
    print " )";
    
    # Search wrap around.
    print " (";
    $replace_fn->($rgn_beg,
                  (defined $pos_wrap_end) ? $pos_wrap_end : $rgn_end,
                  1);
    print " )";
    print "))", "\n";
    print ";;; EOF", "\n";
}

sub main () {
    my $fn_body   = shift @ARGV or die "No input file name!";
    my $fn_out    = shift @ARGV or die "No output file name!";
    my $fn_regx   = shift @ARGV or die "No regexp file name!";
    my $fn_rpla   = @ARGV ? shift @ARGV : die "No replacement file!";
    my $dot_p     = @ARGV ? shift(@ARGV) : die "No dot matches new line flag.";
    my $case_p    = @ARGV ? shift(@ARGV) : die "No case sensitive flag.";
    my $ext_p     = @ARGV ? shift(@ARGV) : die "No extended regular expression flag.";
    my $eval_p    = @ARGV ? shift(@ARGV) : die "No eval replacement flag.";
    my $limit     = @ARGV ? shift(@ARGV) : die "No search limit.";
    my $pos_start = @ARGV ? shift(@ARGV) : die "No start position.";
    my $rgn_beg   = shift @ARGV;
    my $rgn_end   = shift @ARGV;
    
    my $code      = 'utf8';
    
    my($str_body, $str_regx, $str_rpla);
    
    use PerlIO::encoding;
    local $PerlIO::encoding::fallback = Encode::FB_CROAK(); # Die on invalid char.
    {
        local $INPUT_RECORD_SEPARATOR = undef;
        $str_body = FileHandle->new($fn_body, "<:encoding($code)")->getline;
        $str_regx = FileHandle->new($fn_regx, "<:encoding($code)")->getline;
        $str_rpla = $fn_rpla ? FileHandle->new($fn_rpla, "<:encoding($code)")->getline : "";
    }
    
    umask 0177;
    *STDOUT = FileHandle->new($fn_out, ">:encoding($code)");
    
    process_replace(\$str_body, \$str_regx, \$str_rpla,
                    $dot_p, $case_p, $ext_p, $eval_p,
                    length($limit) ? $limit : undef, $pos_start, $rgn_beg, $rgn_end);
    
    exit 0;
}

main();

# EOF

File: bin/foreign-regexp-replace-aux.py
#!/usr/bin/env python

import sys, re, codecs


def interpolate_fn_gn (txt):
    env = {}
    txt = escape_str_for_interpolate_fn_gen(txt)
    
    m_lambda = re.match(r'^lambda', txt)
    m_def = re.search(r'^def\s+([^\s]+)\s*\(', txt, re.MULTILINE)
    
    if (m_lambda):
        txt = "interpolate_fn = " + txt
    elif (m_def):
        txt = (txt + "\ninterpolate_fn = " + m_def.group(1) + "\n")
    else:
        sys.stderr.write('Replacement is not type of lambda or def.')
        exit(1)
    
    try:
        exec(txt, {}, env)
    except Exception as e:
        sys.stderr.write("Error in replacement string: ")
        sys.stderr.write(str(e))
        exit(1)
    
    return env['interpolate_fn']


def escape_str_for_interpolate_fn_gen (txt):
  txt = re.sub(r'"', r'\"', txt)
  return txt


def escape_python_str_for_emacs (txt):
    txt = re.sub(r'\\', r'\\\\', txt)
    txt = re.sub(r'"', r'\\"', txt)
    return txt


pos_wrap_end = None
count_match = 0

def do_replace (f, str_body, regx, str_rpla, interpolate_fn, limit, rgn_beg, rgn_end, wrap_p):
    global pos_wrap_end
    global count_match
    
    for match in regx.finditer(str_body, rgn_beg):
        if (limit != None) and (limit <= count_match):
            break
        
        if match.end() > rgn_end:
            break
        
        if wrap_p and (pos_wrap_end != None) and (pos_wrap_end <= match.start()):
            break
        
        if (not wrap_p) and (pos_wrap_end == None):
            pos_wrap_end = match.start()
        
        f.write("((")
        f.write("%d %d " % match.span())
        for i in range(len(match.groups())):
            if (match.groups())[i] != None: # Skip unmatched group.
                f.write("%d %d " % match.span(i+1))
        f.write(")\"")
        
        try:
            rpla_expanded = (interpolate_fn and
                             [interpolate_fn(match)] or
                             [match.expand(str_rpla)])[0]
            if not (isinstance(rpla_expanded, unicode) or isinstance(rpla_expanded, str)):
                rpla_expanded = str(rpla_expanded)
            
            f.write(escape_python_str_for_emacs(rpla_expanded))
        except Exception as e:
            sys.stderr.write("Error while interpolating replacement: ")
            sys.stderr.write(str(e))
            exit(1)
        
        f.write("\")")
        
        count_match += 1


def main ():
    [prog, fn_body, fn_out, fn_regx, fn_rpla,
     dot_p, case_p, ext_p, eval_p,
     limit, pos_start, rgn_beg, rgn_end] = sys.argv
    
    limit     = (len(limit)     and [int(limit)]     or [None])[0]
    pos_start = (len(pos_start) and [int(pos_start)] or [None])[0]
    rgn_beg   = (len(rgn_beg)   and [int(rgn_beg)]   or [None])[0]
    rgn_end   = (len(rgn_end)   and [int(rgn_end)]   or [None])[0]
    
    str_body = codecs.open(fn_body, 'r', 'utf_8').read()
    str_regx = codecs.open(fn_regx, 'r', 'utf_8').read()
    str_rpla = (fn_rpla and [codecs.open(fn_rpla, 'r', 'utf_8').read()] or [""])[0]
    
    f = codecs.open(fn_out, 'w', 'utf8')
    
    
    try:
        regx = re.compile(str_regx, ((dot_p        and re.DOTALL     or 0) |
                                     ((not case_p) and re.IGNORECASE or 0) |
                                     (ext_p        and re.VERBOSE    or 0) |
                                     re.MULTILINE)) # XXX: Put re.UNICODE flag?
    except Exception as e:
        sys.stderr.write("Error while compiling regexp: ")
        sys.stderr.write(str(e))
        exit(1)
        
    interpolate_fn = (eval_p and [interpolate_fn_gn(str_rpla)] or [None])[0];
    
    rgn_beg   = rgn_beg or 0
    rgn_end   = rgn_end or len(str_body)
    pos_start = ((pos_start < rgn_beg) and
                 [rgn_beg] or
                 [((pos_start > rgn_end) and
                   [rgn_end] or
                   [pos_start])[0]])[0]
    
    f.write("(setq result '((")
    
    do_replace(f, str_body, regx, str_rpla, interpolate_fn, limit, pos_start, rgn_end, False)
    
    f.write(")(")
    
    do_replace(f, str_body, regx, str_rpla, interpolate_fn, limit, rgn_beg,
               ((pos_wrap_end != None) and
                [pos_wrap_end] or
                [rgn_end])[0],
               True)
    
    f.write(")))\n;;; EOF\n")
    
    exit

main()

#EOF
